### ISSUE TITLE
[2023] Mole-BERT: Rethinking Pre-training Graph Neural Networks for Molecules

### ISSUE THREAD

**Issue by mjrs33** [mjrs33/paper#68]
## 概要
VQ-VAEを用いて原子を様々なトークンに割り当てることで得られたembeddingを用いてマスクされた原子の予測をする事前学習に加え、分子の類似度を考慮した対照学習でグラフレベルの事前学習を行うMole-BERTを提案。通常のノードマスクと比較して、同じ原子でも環境に応じた別個のembeddingになる。多くの下流タスクでSOTA性能。

### 論文リンク
https://openreview.net/forum?id=jevY-DtiZTR

### 著者/所属機関
Jun Xia and Chengshuai Zhao and Bozhen Hu and Zhangyang Gao and Cheng Tan and Yue Liu and Siyuan Li and Stan Z. Li

### 投稿日付
2023/2/2

## 背景
BERTの事前学習のように、原子をマスクしてそれを予測する手法（AttrMask）が分子の事前学習として用いられている。しかし、いくつかのケースでnegative transferが起こることが確認されている。

## 新規性
negative transferが起こる原因を解明し、それに対処する。
VQ-VAEを用いて原子のembeddingを得る。そのembeddingをマスクし予測する事前学習を行う（MAM）。
さらに、分子の類似度を考慮した対照学習（TMCL）を提案。
これらを組み合わせたMole-BERTがSOTA性能を示した。

### 既存手法の問題点確認
AttrMaskの事前学習を行った時の学習曲線を見ると、すぐに精度が高くなっており、118という少ない原子数では、この事前学習は簡単すぎることが分かる（左図）。これは語彙が多いBERTの事前学習（MAM）とは対照的である。

![6863326F-EFEE-4346-81E7-E470283B9A18](https://user-images.githubusercontent.com/40561031/218287934-dd675d8f-7d7f-4a06-83b2-49f4452a054f.jpeg)

さらに、分子では原子の多様性が少なく、Cに寄った予測をすることで容易に高精度を実現できてしまう。

NLPはトークナイザーの学習と、言語モデルの事前学習との2ステップとみなすことができる。分子の事前学習はトークンを原子と決め打ちしてしまっており良くない。
著者らは、それぞれの環境で、たとえ同じ種類の原子であったとしても、別々のトークンである必要があると考えた。
例えば、アルデヒドとエーテル内の炭素原子はそれぞれ別の特性を持つため、別物として考えるべきである。

VQ-VAEを用いて、環境に応じた原子embeddingを得るようにした。これにより、上図右のように、それぞれの環境に応じた炭素原子の潜在表現を得ることができた。これはトークン数を多様かつ偏りがないようにする効果がある。
この潜在表現について、ノードレベルのマスクを行い、事前学習を実施した。（MAM: dubbed Masked Atoms Modeling）

また、グラフレベルの対照学習も有効であることが報告されているが、これまでの手法は、別の分子を、分子の類似度に関係なく、同等に離そうと学習が進んでしまう。この問題に対処するため、triplet masked contrastive learning (TMCL)を提案し、分子の類似度を考慮した。

### MAM
全体像は以下。

![DDE17F63-F00E-410A-906C-FA3114867E06](https://user-images.githubusercontent.com/40561031/218287947-b10196dd-6ff5-4468-a23e-0a592659d388.jpeg)

まず、GNNで各ノードが埋め込まれる。次にvector quantizer (VQ)がcodebook内の最も近いembeddingに変換する。これで各原子が用意されたトークンのいずれかに変換されたことになる。このとき、同じ元素であっても複数のトークンがあるため、別のトークンになりうる。

このようにして得られた各ノードのembeddingをGNNのデコーダに流し、元の分子を復元する。VQによって微分不可能になっているが、デコーダからエンコーダに勾配をコピーすることにより学習が可能。なお、学習のlossは再構成誤差に加え、codebookをアップデートするための項、エンコーダのembeddingがcodebookのembeddingに近くなるようにするための項が含まれている。

![43044197-DF62-45E0-84B8-5AE61EC16DDC](https://user-images.githubusercontent.com/40561031/218287949-3b7c7f06-8057-43fc-9910-0a607431d55a.jpeg)

また、通常のVQ-VAEを用いると、異なる元素が同じcodebookのembeddingに行き着く問題が生じたため、グループ化する必要があったそう。
つまり、1-128は炭素、129-256は窒素などのcodebookを作り、そのなかから最も近いembeddingを抽出した。

この学習されたトークナイザーを用いて、ノードを変換する。15%をマスクし、それを予測する事前学習を実施した。

### TMCL
MAMではグラフレベルの特徴を捉えられないが、グラフレベルの事前学習はラベルが必要なことが多い。
対照学習はラベルが不要であるのでそれを利用。

通常の対照学習では、ある分子に変化を与えたものを正例、他の分子を負例とする。しかし、これはアンカー分子と他の分子との類似性を考慮できていない。
TMCLはこの問題に対処するため、アンカー分子から2つの負例を作成する。1つは小さい割合で原子をマスクする(M1)。もうひとつはより大きい割合で原子をマスクする(M2)。M1はM2よりもアンカー分子に近いとみなすことができる。

それぞれのグラフレベルの特徴量ベクトルに対し、
元の分子とM1の類似度　＞　元の分子とM2の類似度
になるように学習を行う。

### データセット
ZINC15 (200万分子)で事前学習を行った。

### モデル
バックボーンモデルとして、5層のGINを用いた。100 epochの事前学習を実施し、100 epochのfine-tuningを実施。 

## 結果
各下流タスクの性能は以下。

![6C6E1978-43ED-4644-8A9B-B19406BDEC3D](https://user-images.githubusercontent.com/40561031/218287953-cc3da40c-ecf4-4594-bbc4-ca64370c376f.jpeg)

AttrMaskで観測されたnegative transferはMAMでは起こらなかった。

Mole-BERTはほとんどのタスクでSOTA性能を記録。また、バックボーンモデルを他のものに変えても効果があることを確認している。

ある分子に対し、データセット内の一致度が高い5件を抽出した結果が以下。

![27FC40FC-05C5-424F-A90B-246C48A74575](https://user-images.githubusercontent.com/40561031/218287958-a288fdf0-6bb8-43fb-972a-e7a3b1b4c519.jpeg)

Mole-BERTはfingerprintベースの一致度が高いものが抽出されている。MAMでは低いものも抽出されているので、MAMだけでは事前学習は不十分で、グラフレベルの特徴量の事前学習が必要であることを示している。
